File: rxbinding/src/main/java/com/jakewharton/rxbinding2/widget/RatingBarRatingChangeObservable.java
package com.jakewharton.rxbinding2.widget;

import android.widget.RatingBar;
import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.MainThreadDisposable;

import static io.reactivex.android.MainThreadDisposable.verifyMainThread;

final class RatingBarRatingChangeObservable extends Observable<Float> {
  private final RatingBar view;

  RatingBarRatingChangeObservable(RatingBar view) {
    this.view = view;
  }

  @Override protected void subscribeActual(Observer<? super Float> observer) {
    verifyMainThread();
    Listener listener = new Listener(view, observer);
    view.setOnRatingBarChangeListener(listener);
    observer.onSubscribe(listener);
    observer.onNext(view.getRating());
  }

  static final class Listener extends MainThreadDisposable
          implements RatingBar.OnRatingBarChangeListener {

    private final RatingBar view;
    private final Observer<? super Float> observer;

    Listener(RatingBar view, Observer<? super Float> observer) {
      this.view = view;
      this.observer = observer;
    }

    @Override public void onRatingChanged(RatingBar ratingBar, float rating, boolean fromUser) {
      if (!isDisposed()) {
        observer.onNext(rating);
      }
    }

    @Override protected void onDispose() {
      view.setOnRatingBarChangeListener(null);
    }
  }
}

File: rxbinding/src/main/java/com/jakewharton/rxbinding4/widget/RatingBarRatingChangeObservable.kt
@file:JvmName("RxRatingBar")
@file:JvmMultifileClass

package com.jakewharton.rxbinding4.widget

import android.widget.RatingBar
import android.widget.RatingBar.OnRatingBarChangeListener
import androidx.annotation.CheckResult
import com.jakewharton.rxbinding4.InitialValueObservable
import io.reactivex.rxjava3.core.Observer
import io.reactivex.rxjava3.android.MainThreadDisposable

import com.jakewharton.rxbinding4.internal.checkMainThread

/**
 * Create an observable of the rating changes on `view`.
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 *
 * *Note:* A value will be emitted immediately on subscribe.
 */
@CheckResult
fun RatingBar.ratingChanges(): InitialValueObservable<Float> {
  return RatingBarRatingChangeObservable(this)
}

private class RatingBarRatingChangeObservable(
  private val view: RatingBar
) : InitialValueObservable<Float>() {

  override fun subscribeListener(observer: Observer<in Float>) {
    if (!checkMainThread(observer)) {
      return
    }
    val listener = Listener(view, observer)
    view.onRatingBarChangeListener = listener
    observer.onSubscribe(listener)
  }

  override val initialValue get() = view.rating

  private class Listener(
    private val view: RatingBar,
    private val observer: Observer<in Float>
  ) : MainThreadDisposable(), OnRatingBarChangeListener {

    override fun onRatingChanged(ratingBar: RatingBar, rating: Float, fromUser: Boolean) {
      if (!isDisposed) {
        observer.onNext(rating)
      }
    }

    override fun onDispose() {
      view.onRatingBarChangeListener = null
    }
  }
}
